src/shared/user-record.c

```diff
@@ -21,6 +23,98 @@
 #define DEFAULT_RATELIMIT_BURST 30
 #define DEFAULT_RATELIMIT_INTERVAL_USEC (1*USEC_PER_MINUTE)
 
+#if ENABLE_COMPAT_MUTABLE_UID_BOUNDARIES
+static int parse_alloc_uid(const char *path, const char *name, const char *t, uid_t *ret_uid) {
+        uid_t uid;
+        int r;
+
+        r = parse_uid(t, &uid);
+        if (r < 0)
+                return log_debug_errno(r, "%s: failed to parse %s %s, ignoring: %m", path, name, t);
+        else if (uid == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "%s: %s == 0, ignoring.", path, name);
+
+        *ret_uid = uid;
+        return 0;
+}
+
+static int read_login_defs(login_defs *ret_defs, const char *path) {
+        _cleanup_fclose_ FILE *f = NULL;
+        uid_t sys_uid_max = SYSTEM_UID_MAX;
+        gid_t sys_gid_max = SYSTEM_GID_MAX;
```

Comment:
maybe make these two an instance of `struct login_defs` here too? no need to allocate these fields separately on the stack, if you can just allocate the whole structure in one, after all you need all its fields here anyway. Also makes the exit path simpler, as you can just copy the whole struct into teh return buffer.